Refactor reconcilers to operate on ClusterResolutionResult

robobario opened the pull request:
### Type of change

- Refactoring

### Description

This extends ClusterResolutionResult to contains a full graph of the attempted resolutions, including the Referent resource if we could obtain it. We need to expose this partial result to generate the ingresses, to try and keep port allocation as stable as we can.

We remove the `Map<LocalRef, ReferentType>` maps from the ProxyResolutionResult as now all clients use ClusterResolutionResult to avoid having to dereference again.

Contributes towards #1920

### Why
Previously the `ProxyResolutionResult` contained referent maps and the client code had to dereference them again. It seemed ugly to have to go back in and out of the API when the DependencyResolver was already dereferencing everything. Why should the clients have to do it again?

The end result is the proxy reconciler does not have to dereference the resources again, it can operate on the ResolvedCluster.

We are about to add functionality to the VirtualKafkaClusterreconciler where we will need to access all referents to compute a checksum from their metadata. Until now we haven't needed to access the referents, so they aren't part of the ClusterResolutionResult, this structure will also make it easy to obtain all the Referents for
checksumming in a subsequent PR.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] PR raised from a fork of this repository and made from a branch rather than main. 
- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

robobario commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,42 +6,57 @@
+                                      List<ResolutionResult<KafkaProtocolFilter>> filterResolutionResults,
+                                      ResolutionResult<KafkaService> serviceResolutionResult,
+                                      List<IngressResolutionResult> ingressResolutionResults,
+                                      ResolutionScope resolutionScope) {
```
This is smelly, ClusterResolutionResult behaves differently in different contexts.

1. When we are resolving the resources in the VKC Reconciler, we do not want to consider the cluster's ResolvedRefs and staleness because we are reconciling that resource.
2. When we are resolving the resources in the Proxy Reconciler, we want to report a stale or ResolvedRefs false VKC as a problem

Maybe this is suggesting it's not a `ClusterResolutionResult` level problem, which could be changed to be purely about the VKC's Referents.

Modelling it on some object associated with the cluster feels right, like I don't want to ask the `ProxyResolutionResult` `isClusterStale(cluster)`. Maybe it could be a type hierarchy instead so that there's a ProxyClusterResolutionResult that wraps the referent one and adds these modifications.

robobario replied:
after a chat with Sam I'm going to change it from being state of the ClusterResolutionResult to adding more explicit methods to the result so each context will be asking it for the problems it cares about

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/model/ingress/Ingresses.java:
```diff
@@ -12,11 +12,12 @@
-import io.kroxylicious.kubernetes.operator.resolver.ProxyResolutionResult;
+import io.kroxylicious.kubernetes.operator.resolver.ClusterResolutionResult;
 
 class Ingresses {
```
Please can we have Javadoc that explains what an `Ingresses` is, conceptually speaking?

We have:
* IngressDefinition
* IngressInstance
* Ingresses
* ProxyIngressModel

I have a real hard time understanding what's going on there.

robobario replied:
I don't want to go too far down tidying the ingress of things up as it's a mess. I'll rename `Ingresses` to `IngressDefinitionBuilder` in this PR to give it a better name and disambiguate from the `Ingresses` class generated (with the annoying pluralization) from the CR.

robobario replied:
ah sorry, you were requesting javadoc, I'll have a shot at improving it

tombentley replied:
Thanks! I hadn't even clocked that we had two `Ingresses`. Definitely distinct names would be good.

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/model/ProxyModel.java:
```diff
@@ -22,6 +21,6 @@
  */
 public record ProxyModel(ProxyResolutionResult resolutionResult,
```
Is a `ProxyModel` really a "`ResolvedProxy`"?

robobario replied:
I'm using `Resolved` to mean 100% of the recursive dependencies could be resolved (putting aside the fact it doesn't consider Secrets/Configmaps yet). 

`ProxyModel` contains information about problems/unresolved referents in the ProxyResolutionResult.

Maybe like you're getting at later we should have everything operate on the `*ResolutionResult` and drop the `Resolved*` conveniences.

robobario replied:
have removed the `Resolved*` classes

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
+ * @param serviceResolutionResult the resolution result for service referenced by the cluster
+ * @param ingressResolutionResults the resolution results for ingresses referenced by the cluster, in the same order they are declared in the VirtualKafkaCluster spec
  */
 public record ClusterResolutionResult(VirtualKafkaCluster cluster,
```
What about Secrets, ConfigMaps referenced from a `VirtualKafkaCluster.spec.ingresses.tls` or `KafkaService.spec.tls`? If they're really not required then explain why in the Javadoc.

robobario replied:
Yeah it would make sense to resolve them here I think, as a dangling ref to a secret encountered in the aggregator should be reported back as a problem to the VKC. It just hasn't been done yet, so I'd be keen to leave it for another PR.

k-wall replied:
+1.I started down that path when I did the the virtualcluster TLS work.  The PR got unwieldy and I stepped back to something smaller.

One unsolved issue I had was where to implement the Secret and ConfigMap content validations. By that I mean, the Kubernetes Secret has to be of type `kubernetes.io/tls or a particular data key has to end with `.pem`, `.p12` etc). Either the ClusterResolutionResult needs to expose the successfully resolved resources (so that the validation can be done in the Reconcilers), or the DependencyResolver needs to accept arbitrary validators (and have a mechanism to get back sufficient details to write a good condition message).   The former seemed more appealing.

robobario replied:
+1, let resolver handle lookups, referential integrity (including considering the ResolvedRefs conditions of Referents), stale status check also feels closely related.

But once we are looking into the Referents, inspecting their spec/data I think that's some other component's job.

SamBarker replied:
> But once we are looking into the Referents, inspecting their spec/data (beyond following refs) I think that's some other component's job.

Yeah I think that model works nicely. For adding checksums to Virtual clusters I was looking for something to give me the fully resolved and valid graph that I could just walk but it doesn't really exist on main and is a quite awkward to extract as things stand.

robobario replied:
Yep, after this branch merges it should be easy to add a `Stream<HasMetadata> ClusterResolutionResult#referents()`

robobario replied:
Or you can walk the different Lists of Referents if the client wants more control over order.

robobario replied:
> fully resolved and valid graph

I had this idea implemented earlier in this PR, but backed off it because there's already a lot of change/noise and classes to understand (ResolveCluster vs VirtualKafkaCluster vs ClusterResolutionResult). Can revisit again in a future PR.

robobario replied:
created #2169

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
+     * All referents and cluster are fully resolved iff
+     * <ol>
+     *     <li>All references can be retrieved from kubernetes (we have no dangling refs)</li>
+     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has unresolved dependencies</li>
```
And the condition is up to date?

robobario replied:
good point, I don't think our code considers that currently.

robobario replied:
created #2146

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
+    }
+
+    /**
+     * Transforms this result into a ResolvedCluster, it is assumed that the client has checked that this ClusterResolutionResult
```
"Transforms this result into a ResolvedCluster"

I think you need to explain somewhere why you need a separate representations of the result from the resolved things.

robobario replied:
removed `Resolved*` classes

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
-        Objects.requireNonNull(from);
-        Objects.requireNonNull(kindTo);
-        return danglingReferences.stream().filter(r -> r.from().equals(from) && kindTo.equals(r.to.getKind())).map(DanglingReference::to);
+    public Stream<LocalRef<?>> findDanglingReferences(LocalRef<?> fromReferrer, String toKind) {
```
It's very confusing to overload a method with one that has a different result type.

robobario replied:
fixed, they all return `Stream<DanglingReference>` now

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
-     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has unresolved dependencies</li>
-     * </ol>
+     * Find referents and cluster with stale status (status.observedGeneration != metadata.generation)
+     * @return LocalRefs to referents and cluster with stale status
```
I don't understand what "referents and cluster" means, given this returns some `LocalRefs` (==referents). Where does the cluster part come from?

robobario replied:
`referents` means referents of the cluster. the cluster itself isn't a Referent but we want to treat it the same way sometimes. The ProxyReconciler want to ask "do any of the referents or the cluster have ResolvedRefs=false?". Where the VKC Reconciler only wants to ask "do any of the referents have ResolvedRefs=false?"

tombentley replied:
> we want to treat it the same way sometimes

That is not at all obvious, and is unintuitive for anyone with the CR model in their head (where clusters do not refer to clusters).
* Is it really necessary? I mean, can't the `ProxyReconciler` just make two calls? Or could it concatenate a stream of referents with a LocalRef for the cluster at the call site which want to tread things that way?
* If it is really necessary then you should at least explain this in the Javadoc.

robobario replied:
Okay, I'll shuffle it around so we expose a couple of boolean methods and have the clients handle the concatenation

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,197 @@
-     * Describes the case where an entity A references an entity B but we can not find B.
+     * Describes the case where an entity A references an entity B, but we can not find B.
      * @param from
      * @param to
```
Maybe rename `to` to `missing` or `absent`? I suspect that would make call sites more readable.

robobario replied:
done, `to` should be `referrer` as well.

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/IngressResolutionResult.java:
```diff
@@ -0,0 +1,31 @@
+ * @param ingressResolutionResult the dereferenced KafkaProxyIngress result
+ * @param proxyResolutionResult the dereferenced KafkaProxy result
+ */
+public record IngressResolutionResult(Ingresses ingress,
```
I don't know what to make of the plural type but singular parameter name.

tombentley replied:
Again, I don't really understand why the need for a `Result` thing and a `Resolved` thing.

robobario replied:
`Ingresses` is generated code from the CRD, I think this record-field is unused.

robobario replied:
have removed the `Resolved` classes

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ProxyResolutionResult.java:
```diff
@@ -32,26 +24,10 @@
  * describes which dependencies could not be resolved per VirtualKafkaCluster.
```
The comment begins: 

> The result of a deep resolution of the dependencies of a single KafkaProxy.

I think this might be more consistent with the terminology we're using elsewhere: 

> The result of a deep resolution of the transitive references from a single KafkaProxy.

I note that you're using both "resolution" and "dereference". We should try to stick to one.

robobario replied:
purged all `deref` mentions and removed most usages of `dependency/dependencies`. I'll put up a subsequent PR to rename `DependencyResolver` to try and get away from the `dependency` noun altogether.

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/Referent.java:
```diff
@@ -0,0 +1,70 @@
+import io.kroxylicious.kubernetes.operator.ResourcesUtil;
+
+/**
+ * A single referent
```
* This is really modelling the state of all the references from the referrer, I think? If so, perhaps there's a better name for this type.
* Could you give an example? I think that would make it much more concrete for the reader.

robobario replied:
Referent is a little wrapper class around the Referent resource so there is a common interface for finding stale or ResolvedRefs=False Referents. Ideally we'd be able to have these interfaces directly on the Resource classes.

Maybe calling it a `Referent` isn't that useful, it's describing something about what status/conditions the resource has? HasCommonStatus? KroxyliciousResource? Something saying it has some features common to all our CRs.

robobario replied:
Have removed `Referent`, instead the user can get at the resource directly and ask if it's stale/hasResolvedRefsFalseCondition via static methods on ResourcesUtil

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ResolutionResult.java:
```diff
@@ -0,0 +1,43 @@
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * The result of attempting to dereference a reference
```
```suggestion
 * The result of attempting to resolve a reference
```

tombentley commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ResolutionResult.java:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * The result of attempting to dereference a reference
+ * @param referrer the referring resource's LocalRef
```
```suggestion
 * @param referrer the referring resource represented as a LocalRef
```

robobario commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,95 @@
+     */
+    public Stream<HasMetadata> allResolvedReferents() {
+        return allResolutionResults()
+                .filter(resolutionResult -> !Objects.equals(resolutionResult.reference().getKind(), HasMetadata.getKind(KafkaProxy.class)))
```
This is the ugly code I mentioned in standup, `allResolutionResults` contains results for anything that refers to a KafkaProxy, so we can find dangling refs where we could not resolve a KafkaProxy.

Until now we have not considered the staleness of the KafkaProxy resources, I was unsure if we should require a fresh status on all KafkaProxy referents in case we create a deadlock where both the KafkaProxy requires a fresh status on the VKC, and the VKC requires a fresh status on the KafkaProxy.

robobario replied:
would suggest we merge without an answer, keeping the current behaviour, while we discuss

SamBarker replied:
Setting aside the question of wether we should ignore Staleness of KafkaProxy for a moment. It feels strange that this type specific behaviour is buried in a generic class. Personally I think we should overload with a version that accepts the filter and ask the VKC Reconciler to pass its special sauce predicate. The default predicate would just accept everything. 

Should the KafkaProxy Reconciler have the reverse condition to break the cycle on that side too?

tombentley replied:
I think we can wave this away by chanting "eventual consistency" can't we? By which I mean we only care about staleness for resources which have refs of some kind. A `KafkaProxy` does not have any refs, so the KPRec shouldn't care that a VKC which references the KP is stale (KPRec should ignore the staleness and proceed as if if the VKC is fresh). That's because we'd expect the VKCRec to shortly reconcile the VKC, at which point the KP should be reconciled again.

Maybe that's what you were describing when I interrupted during standup, in which case sorry if I derailed things.

robobario replied:
> VKC Reconciler to pass its special sauce predicate. The default predicate would just accept everything.

Both reconcilers do the same resolution work per VKC, resolving proxyRef on VKC and ingresses. It could be possible during KafkaProxy reconcilation that we observe an ingress that refers to a non-existent KafkaProxy and that should cause some failure. The difference is the source of the KafkaProxy, VKC Reconciler gets it from the JOSDK Context, KPReconciler just passes through the primary.

Both reconcilers currently want to exclude these resolved KafkaProxies from staleness checking

robobario replied:
@SamBarker is that a blocker for you? Maybe we could use a better term than `allReferents` maybe it's more accurately the `directReferents` of the cluster? Or we should't offer this stream of all referents munged together and clients should be forced to operate on the referents they are interested in. 

The hassle with that is that there are several places we just want to check that everything is resolved and we are able to proceed, so a bit of duplication required, or a shared static method, and at that point why not have the method on this result class.

SamBarker replied:
No not a blocker for me.

Definitely a clean up we can deal with separately.

robobario commented:
I've deleted `Referent` and removed most of the methods on `ClusterResolutionResult` in favour of offering `Predicate`s for use with the streaming API to filter DanglingReferences and referent resources. Since a lot of the methods were eventually just conveniences the streams API offers anyway like `anyMatch`, `allMatch` etc.

robobario commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,172 @@
+     * @param toKind a Referent kind
+     * @return a stream of all dangling references from the specified LocalRef to the specified kind
+     */
+    public Stream<ClusterResolutionResult.DanglingReference> findDanglingReferences(LocalRef<?> fromReferrer, String toKind) {
```
I've reduced it down to a small amount of methods plus some static Predicates since it was all just conveniences around the streams API. So now clients would call `result.allDanglingReferences().filter(predicate)`

robobario commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ResolutionResult.java:
```diff
@@ -0,0 +1,43 @@
+
+    public T referentResource() {
+        if (referent == null) {
+            throw new IllegalStateException("Referent is null!");
```
yeah NPE is appropriate here

robobario commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/model/ingress/IngressDefinitionBuilder.java:
```diff
@@ -0,0 +1,61 @@
+ * IngressDefinitionBuilder is responsible for creating richer IngressDefinition objects from the raw
+ * Custom Resources that define it.
+ */
+class IngressDefinitionBuilder {
```
This is `Ingresses` renamed, the diff shows it as a create/delete

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/VirtualKafkaClusterReconciler.java:
```diff
@@ -258,28 +261,43 @@ private Optional<OwnerReference> extractOwnerRefFromKubernetesService(Service se
                     joiningMessages(proxyMsg, serviceMsg, ingressMsg, filterMsg));
         }
-        else if (clusterResolutionResult.anyResolvedRefsConditionsFalse() || !unresolvedIngressProxies.isEmpty()) {
+        else if (clusterResolutionResult.allResolvedReferents().anyMatch(ResourcesUtil::hasResolvedRefsFalseCondition) || !unresolvedIngressProxies.isEmpty()) {
```
Again I'd like a method to wrap this compound condition to give it an intention revealing name.

SamBarker commented on kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java:
```diff
@@ -6,109 +6,95 @@
      * <ol>
-     *     <li>All references can be retrieved from kubernetes (we have no dangling refs)</li>
-     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has unresolved dependencies</li>
+     *     <li>All references can be retrieved from kubernetes (we have no dangling references)</li>
```
is it clear to a reader what a `dangling reference` is?

SamBarker reviewed (approved):
A bunch of nits and stylistic thoughts but I think this could be merged as is and the tidying up happen out of band.
